xl-v1-metadata.go

```diff
@@ -259,39 +282,18 @@ func (xl xlObjects) renameXLMetadata(srcBucket, srcPrefix, dstBucket, dstPrefix
 	// Wait for all the routines.
 	wg.Wait()
 
-	// Gather err count.
-	var errCount = 0
-	for _, err := range mErrs {
-		if err == nil {
-			continue
+	if !isQuorum(mErrs, xl.writeQuorum) {
+		if isQuorum(mErrs, xl.readQuorum) {
+			return nil
 		}
-		errCount++
+		xl.undoRenameXLMetadata(srcBucket, srcPrefix, dstBucket, dstPrefix, mErrs)
+		return errXLWriteQuorum
 	}
-	// We can safely allow RenameFile errors up to len(xl.storageDisks) - xl.writeQuorum
-	// otherwise return failure. Cleanup successful renames.
-	if errCount > len(xl.storageDisks)-xl.writeQuorum {
-		// Check we have successful read quorum.
-		if errCount <= len(xl.storageDisks)-xl.readQuorum {
-			return nil // Return success.
-		} // else - failed to acquire read quorum.
-
-		// Undo rename `xl.json` on disks where RenameFile succeeded.
-		for index, disk := range xl.storageDisks {
-			if disk == nil {
-				continue
-			}
-			// Undo rename object in parallel.
-			wg.Add(1)
-			go func(index int, disk StorageAPI) {
-				defer wg.Done()
-				if mErrs[index] != nil {
-					return
-				}
-				_ = disk.RenameFile(dstBucket, dstJSONFile, srcBucket, srcJSONFile)
-			}(index, disk)
+	for _, err := range mErrs {
+		if err != nil && err != errDiskNotFound {
+			xl.undoRenameXLMetadata(srcBucket, srcPrefix, dstBucket, dstPrefix, mErrs)
```

Comment:
doing undoRenameXLMetadata during complete-multipart-upload will be harmful because it will remove xl.json from the .minio/multipart/object/ directory